Patch/extend osw output

jcharkow opened the pull request:
## Description

Extend output in .osw file to include metrics on peak shape on the transition level. Addressing issue https://github.com/OpenMS/OpenMS/issues/7222

## Checklist
- [X ] Make sure that you are listed in the AUTHORS file
- [ X] Add relevant changes and new features to the CHANGELOG file
- [X ] I have commented my code, particularly in hard-to-understand areas
- [ X] New and existing unit tests pass locally with my changes
- [ X] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

jcharkow committed:
[FEATURE] add OSW output

Add output of RT FWHM and mass_error ppm to the feature transition table

jcharkow committed:
[FEATURE] add peak metric outputs to .osw file

add peak metric computations to the .osw FEATURE_TRANSITION file. I
believe these will be the same across transitions for consensus peak
picking however may differ without consensus peak picking?

jcharkow committed:
update changelog

jcharkow committed:
Remove debug lines

timosachsenberg commented:
Nice! As we don't have integration tests, I don't know if this could affect downstream processing tools of OpenSWATH negatively. But I guess you have a much better picture here ...

jcharkow commented:
Downstream processing tools like pyprophet only use columns that start with `var` so it should not affect pyprophet results.

timosachsenberg commented on src/openms/source/ANALYSIS/OPENSWATH/OpenSwathOSWWriter.cpp:
```diff
@@ -211,23 +228,61 @@ namespace OpenMS
-      for (const auto& sub_it : feature_it.getSubordinates())
+      std::vector<String> masserror_ppm = feature_it.metaValueExists("masserror_ppm") ? getSeparateScore(feature_it, "masserror_ppm") : std::vector<String>();
+
+      auto subordinates = feature_it.getSubordinates();
```
const auto& might prevent a copy here

timosachsenberg commented on src/openms/source/ANALYSIS/OPENSWATH/OpenSwathOSWWriter.cpp:
```diff
@@ -211,23 +228,61 @@ namespace OpenMS
+      auto subordinates = feature_it.getSubordinates();
+      for (Size i=0; i < subordinates.size(); i++)
       {
+        auto sub_it = subordinates[i];
```
does const auto& work here?

jcharkow replied:
yes

jcharkow committed:
Apply suggestions (add const auto)

cbielow commented on src/openms/include/OpenMS/ANALYSIS/OPENSWATH/OpenSwathOSWWriter.h:
```diff
@@ -103,6 +104,7 @@ namespace OpenMS
                        const String& input_filename = "inputfile",
                        bool ms1_scores = false,
                        bool sonar = false,
+                       bool compute_peak_shape = false,
```
the interface is very easy to use wrong (in general), since its just a bunch of bools -- easy to get the order wrong.
In this particular case its even more dangerous, since old code with
```
OpenSwathWriter("bla", 0, "myfile", true, false, true)
```
still works after this change, but does something completely different!

Suggestions:
1) replace each boolean with an ENUM, e.g. `enum class SONAR {ON, OFF}` which results in `..., SONAR:ON, ...` etc

Use a struct/class which contains the values and allows to set proper names via chaining
```
myParams().sonar(false).computePeakShape(true). ...
```

timosachsenberg replied:
would chaining work with pyopenms?

cbielow replied:
not a pyOpenMS expert, but each functions simply returns a reference to self.
We should not return a copy here, since the changes will be lost unless assigned to a final result object, i.e.
`OpenSwathWriter("bla", 0, "myfile", myParams().sonar(false).computePeakShape(true))` only works when `MyParam& sonar(bool state)` returns a reference, not a copy.

timosachsenberg replied:
@jpfeuffer I always get confused with how temporaries are handled in pyOpenMS. Would in principle something like @cbielow proposed work?

jpfeuffer replied:
I think the default wrapping for pyopenms is to always make a copy. You will need a custom pyx or a new wrap-statement to achieve that. (e.g. wrap-return-self where you do something like

```
self.inst.get().function()
return self
```
)

jpfeuffer replied:
Would be kinda nice indeed

jcharkow replied:
For now I was thinking of just implementing the first suggestion e.g. enum class SONAR {ON, OFF} to not give us trouble with pyopnems

timosachsenberg replied:
hmm better but using the enum is here also a bit non-standard.
Without looking at the details, could another option be to use the OpenMS DefaultParamHandler?
e.g. similar to here: https://github.com/OpenMS/OpenMS/blob/57817d632cf70d45eeb52817c3f6afc2f54921f4/src/openms/source/FORMAT/MSPFile.cpp

jcharkow replied:
We could however I am not sure if this is a bit overkill for this application? Theoretically, none of these parameters should be set manually they should all be read off of the parameters set with OpenSwathWorkflow? I am not sure if there is a way for the class to just determine its own parameters based on the feature it is provided? E.g. look for SONAR data and if it is found than initiate with SONAR?

timosachsenberg replied:
If it is only set in OpenSWATHWorkflow and not exposed to python then it is of course ok

jcharkow replied:
So it is exposed to python but automatic handling might still work? I am going to look into it further

cbielow commented on src/topp/OpenSwathWorkflow.cpp:
```diff
@@ -921,7 +923,7 @@ class TOPPOpenSwathWorkflow
     FeatureMap out_featureFile;
     OpenSwathTSVWriter tsvwriter(out_tsv, file_list[0], use_ms1_traces, sonar); // only active if filename not empty
-    OpenSwathOSWWriter oswwriter(out_osw, run_id, file_list[0], use_ms1_traces, sonar, enable_uis_scoring); // only active if filename not empty
+    OpenSwathOSWWriter oswwriter(out_osw, run_id, file_list[0], use_ms1_traces, sonar, compute_peak_shape_metrics, enable_uis_scoring); // only active if filename not empty
```
there should  be a test which covers the new functionality!

jcharkow replied:
tests are a bit difficult because I do not know if the output for .osw can be really compared. I can add a test with compute_peak_shape_metrics to the black-box tests but I do not think there are currently any tests for OpenSwathOSWWriter

timosachsenberg replied:
That would be great

jcharkow committed:
Merge branch 'develop' of github.com:OpenMS/OpenMS into patch/extend_osw_output

jcharkow committed:
minor: code style changes

jcharkow committed:
Add apex_RT to output

jcharkow committed:
Add test for peak shape metrics

jcharkow committed:
fix: avoid division by 0 in peak shape calculations

Avoid division by 0 which can occur if the start width and start_5 (or
start_10) are the same point. This can cause a crash

jcharkow committed:
feature: auto set adding compute_peak_metrics

automatically add to output if the feature is present.

jcharkow committed:
feature: automatically determine if sonar

Also add tests with SONAR output part of the .osw file

jcharkow committed:
feature: autodetermine if ms1 is used

jcharkow committed:
add compute shape metrics to UIS scores

jcharkow committed:
fix: enable_ms1 flag bug.

use different score to check if ms1 exists since intensity can sometimes
be 0

jcharkow commented:
In order to make the parameters less confusing I added a way to automatically determine if sonar, peak shape scores and ms1 are found so do not have to pass parameters for them.

jcharkow commented:
@singjc Can you please test that this works as expected with UIS? (e.g. no data is missing from the .osw file)?

jcharkow committed:
fix: ppm_diff reporting

bug that could get segmentation fault when recording ppm differences
because would not report the diffrences if there is no signal found.
Change this behaviour that report -1 if no signal found to avoid
segmentation fault.

jcharkow committed:
test: update tests for new patch

update tests to reflect the new masserror_ppm scoring behaviour

jcharkow committed:
Merge branch 'develop' into patch/extend_osw_output

jcharkow committed:
fix: bug in tests

singjc commented:
> @singjc Can you please test that this works as expected with UIS? (e.g. no data is missing from the .osw file)?

I just got around to testing this, and there seems to be some SQL write issue when writing out the RT_FWHM to the FEATURE_TRANSITION table.

```
error message: NOT NULL constraint failed: FEATURE_TRANSITION.RT_FWHM
```

I think one (some) of the identification transitions end up with an invalid FWHM, I will look a bit more into where this occurs.

jcharkow committed:
fix: bug in UIS code

UIS RT FWHM was not recording properly, also realized that transitions
in UIS may have not been recording properly either.

jcharkow committed:
bug fixes for UIS

jcharkow commented:
@timosachsenberg I think this is ready for merging? I have not had a chance to review with @hroest but this does not change any functionality just outputs more information so I think he would be ok with it. @singjc also tested and does not have any errors

timosachsenberg commented on src/tests/topp/FeatureFinderIdentification_1_output.featureXML:
```diff
@@ -377,7 +377,7 @@
 			<UserParam type="float" name="rightWidth" value="2197.34326171875"/>
 			<UserParam type="float" name="peak_apices_sum" value="4.861266306152344e05"/>
-			<UserParam type="floatList" name="masserror_ppm" value="[456.256957954320967, -42.506989184944054]"/>
+			<UserParam type="floatList" name="masserror_ppm" value="[-1.0, -42.506989184944054]"/>
```
Using -1 as magic constant could be a bit dangerous. Is this properly documented?

jcharkow replied:
Typically this score would not be used in downstream analysis tools but rather just for the users information. I will add documentation in the function.

jcharkow committed:
update documentation for changed scoring convention

cbielow commented on src/openms/include/OpenMS/ANALYSIS/OPENSWATH/DIAScoring.h:
```diff
@@ -41,8 +41,11 @@ namespace OpenMS
 
-    - massdiff score: computes the difference in ppm of the experimental signal
-         to the expected signal (thus less is better)
+    - massdiff score: computes the difference in ppm of the experimental signal. 
+         to the expected signal (thus less is better). 
+         sum(ppm_difference) / # transitions
+         Divide by the total number of transitions and is thus quite punnishing if a transition is missing
+         Also outputs a list of all the ppm differences, if signal is not found output -1.0
```
try to build the OpenMS docs locally, and inspect what Doxygen makes of this. It will ignore manual linebreaks, so this will read differently. May need some adjustments.

cbielow commented on src/openms/include/OpenMS/ANALYSIS/OPENSWATH/OpenSwathScores.h:
```diff
@@ -182,11 +183,28 @@ namespace OpenMS
     std::vector<double> ind_mi_score;
+    
+    // peak shape metrics
+    std::vector<double> ind_start_position_at_5;
+    std::vector<double> ind_end_position_at_5;
+    std::vector<double> ind_start_position_at_10;
+    std::vector<double> ind_end_position_at_10;
+    std::vector<double> ind_start_position_at_50;
+    std::vector<double> ind_end_position_at_50;
+    std::vector<double> ind_total_width;
+    std::vector<double> ind_tailing_factor;
+    std::vector<double> ind_asymmetry_factor;
+    std::vector<double> ind_slope_of_baseline;
+    std::vector<double> ind_baseline_delta_2_height;
+    std::vector<double> ind_points_across_baseline;
+    std::vector<double> ind_points_across_half_height;
```
quite a few more datastructures here. I assume the memory footprint of OpenSwathWorkflow will not become unbearable?

jcharkow replied:
It should be fine because usually there are only 6 entries for each at a time (because usually run with 6 transitions). Also I don't expect this functionality to be used except for advanced use cases.

cbielow commented on src/openms/source/ANALYSIS/OPENSWATH/MRMFeatureFinderScoring.cpp:
```diff
@@ -349,11 +350,28 @@ namespace OpenMS
       std::vector<double> ind_intensity_ratio;
       std::vector<double> ind_mi_ratio;
 
+      // peak shape metrics
+      std::vector<double> ind_start_position_at_5;
+      std::vector<double> ind_end_position_at_5;
+      std::vector<double> ind_start_position_at_10;
+      std::vector<double> ind_end_position_at_10;
+      std::vector<double> ind_start_position_at_50;
+      std::vector<double> ind_end_position_at_50;
+      std::vector<double> ind_total_width;
+      std::vector<double> ind_tailing_factor;
+      std::vector<double> ind_asymmetry_factor;
+      std::vector<double> ind_slope_of_baseline;
+      std::vector<double> ind_baseline_delta_2_height;
+      std::vector<double> ind_points_across_baseline;
+      std::vector<double> ind_points_across_half_height;
```
this looks like an exact duplicate of `struct OPENMS_DLLAPI OpenSwath_Ind_Scores`?!
Above you already instanciate `idscores`. Why not work on that object directly? Would save a lot of code.

cbielow commented on src/topp/OpenSwathWorkflow.cpp:
```diff
@@ -774,6 +775,12 @@ class TOPPOpenSwathWorkflow
       feature_finder_param.setValue("Scores:use_uis_scores", "true");
     }
 
+    bool compute_peak_shape_metrics = feature_finder_param.getValue("TransitionGroupPicker:compute_peak_shape_metrics") == "true";
```
```suggestion
    bool compute_peak_shape_metrics = feature_finder_param.getValue("TransitionGroupPicker:compute_peak_shape_metrics").toBool();
```

jcharkow committed:
apply cbielow suggestions

cbielow reviewed (approved):
nice!
Thanks for the fixes.

Good to merge.
